{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **RADI605: Modern Machine Learning**\n",
    "\n",
    "### Assignment: Gradient Boosting\n",
    "**Romen Samuel Rodis Wabina** <br>\n",
    "Student, PhD Data Science in Healthcare and Clinical Informatics <br>\n",
    "Clinical Epidemiology and Biostatistics, Faculty of Medicine (Ramathibodi Hospital) <br>\n",
    "Mahidol University\n",
    "\n",
    "Note: In case of Python Markdown errors, you may access the assignment through this GitHub [Link](https://github.com/rrwabina/RADI605)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.  From the Gene Expression Microarray Data with 4 genes and 2 classes <code>(Cancer=[Yes, No])</code>. Create a Gradient Boost Tree by showing step-by-step calculation and setting $M = 1$, $\\alpha = 0.3$ and <code>max depth = 3</code>.\n",
    "\n",
    "<center>\n",
    "<img src = \"../figures/fig.PNG\" width = \"450\"/> <br>\n",
    "</center>\n",
    "\n",
    "<!-- | Samples   | Gene 1 | Gene 2 | Gene 3 | Gene 4 | Cancer | \n",
    "|:---------:|--------|--------|--------|--------|--------|\n",
    "|     1     |  4.80  |  3.40  |  1.90  |  0.20  |  Yes   | \n",
    "|     2     |  5.00  |  3.00  |  1.60  |  1.20  |  Yes   | \n",
    "|     3     |  5.00  |  3.40  |  1.60  |  0.20  |  Yes   | \n",
    "|     4     |  5.20  |  3.50  |  1.50  |  0.20  |  Yes   | \n",
    "|     5     |  4.80  |  3.10  |  1.60  |  0.20  |  Yes   | \n",
    "|     6     |  5.40  |  3.40  |  1.50  |  0.40  |  Yes   | \n",
    "|     7     |  6.40  |  3.20  |  4.70  |  1.50  |  No    | \n",
    "|     8     |  6.90  |  3.10  |  4.90  |  1.50  |  No    | \n",
    "|     9     |  5.50  |  2.30  |  4.00  |  1.30  |  No    | \n",
    "|    10     |  6.50  |  2.80  |  4.60  |  1.50  |  No    | \n",
    "|    11     |  5.70  |  2.80  |  4.50  |  1.30  |  No    | \n",
    "|    12     |  6.30  |  3.30  |  4.70  |  1.60  |  No    |  -->"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Solution:\n",
    "We start with a leaf that represents an initial prediction for every individual using <code>log(odds)</code>. Since there are 6 patients in the dataset who have cancer while 6 patients do not have cancer, then the <code>log(odds)</code> that somehave have cancer is\n",
    "$$ \\log(\\frac{6}{6}) = \\log\\left({\\frac{p}{1-p}}\\right) = \\log(\\frac{0.5}{0.5}) = \\log(1) = 0$$\n",
    "From the <code>log(odds)</code>, the probability of having cancer is:\n",
    "$$ \\frac{e^{\\log(\\frac{6}{6})}}{1 + e^{\\log(\\frac{6}{6})}} = \\frac{e^{0}}{1 + e^{0}} = 0.5$$\n",
    "\n",
    "Since the probability of having cancer is greater than or equal to 0.5, we can classify the patients in the dataset as someone who have cancer. We can measure how bad the initial prediction is by calculating psuedo-residuals, the difference between the observed and the predicted values, i.e., <code>residual = observed - predicted</code>.\n",
    "\n",
    "<center>\n",
    "<img src = \"../figures/fig1.PNG\" width = \"750\"/> <br>\n",
    "</center>"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have the <code>residuals</code>, we can build the new tree fomr Gene 1, Gene 2, Gene 3, and Gene 4. We set-up the tree using the given parameters: $M = 1, \\alpha = 0.3,$ and <code>max_depth = 3</code>. To build the tree, the initial proportion of the binary predictions is as follows: $\\frac{6}{12} = 0.5$ for the patients with Cancer while $0.5$ for non-cancer patients. The gini impurity is then\n",
    "\n",
    "\\begin{equation*}\n",
    "\\begin{aligned}\n",
    "    \\text{Gini}(D_1)    &= 1 - \\sum_{j = 1}^{L}p^2_j \\\\\n",
    "                        &= 1 - (0.5)^2 - (0.5)^2 \\\\\n",
    "                        &= 1 - 0.50 \\\\\n",
    "    \\text{Gini}(D_1)    &= 0.50\n",
    "\\end{aligned}\n",
    "\\end{equation*}\n",
    "\n",
    "Let's try to split the data based on Gene 1. We first sort the values in ascending order:\n",
    "$$ [4.8, 4.8, 5.0, 5.0, 5.2, 5.4, 5.5, 5.7, 6.3, 6.4, 6.5, 6.9] $$\n",
    "We then evaluate the sorted values though <code>Gini Gain</code> where $\\text{Gini}(D_l)$ and $\\text{Gini}(D_r)$ is the gini indices for left and right, respectively."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\\begin{equation*}\n",
    "\\begin{aligned}\n",
    "    \\text{GiniGain}(\\text{Gene 1}, 5.0) &= \\left( \\frac{|D_l|}{|D_t|} \\text{Gini}(D_l) + \\frac{|D_r|}{|D_t|} \\text{Gini}(D_r) \\right)  &\n",
    "    \\text{GiniGain}(\\text{Gene 1}, 5.2) &= \\left( \\frac{|D_l|}{|D_t|} \\text{Gini}(D_l) + \\frac{|D_r|}{|D_t|} \\text{Gini}(D_r) \\right)  \\\\\n",
    "    \n",
    "                                        &= \\left( \\left( \\frac{2}{12}  * 0    \\right) + \\left( \\frac{10}{12} * 0.48 \\right)   \\right)  &\n",
    "                                        &= \\left( \\left( \\frac{4}{12}  * 0    \\right) + \\left( \\frac{8}{12}  * 0.38 \\right)   \\right)  \\\\\n",
    "    \\text{GiniGain}(\\text{Gene 1}, 5.0) &= 0.40 & \n",
    "    \\text{GiniGain}(\\text{Gene 1}, 5.2) &= 0.25\n",
    "\\end{aligned}\n",
    "\\end{equation*}\n",
    "\n",
    "\n",
    "\\begin{equation*}\n",
    "\\begin{aligned}\n",
    "    \\text{GiniGain}(\\text{Gene 1}, 5.4) &= \\left( \\frac{|D_l|}{|D_t|} \\text{Gini}(D_l) + \\frac{|D_r|}{|D_t|} \\text{Gini}(D_r) \\right)  &\n",
    "    \\text{GiniGain}(\\text{Gene 1}, 5.5) &= \\left( \\frac{|D_l|}{|D_t|} \\text{Gini}(D_l) + \\frac{|D_r|}{|D_t|} \\text{Gini}(D_r) \\right)  \\\\\n",
    "    \n",
    "                                        &= \\left( \\left( \\frac{5}{12}  * 0.00  \\right) + \\left( \\frac{7}{12}  * 0.24 \\right)   \\right)  &\n",
    "                                        &= \\left( \\left( \\frac{6}{12}  * 0.00  \\right) + \\left( \\frac{8}{12}  * 0.00 \\right)   \\right)  \\\\\n",
    "    \\text{GiniGain}(\\text{Gene 1}, 5.4) &= 0.14 & \n",
    "    \\text{GiniGain}(\\text{Gene 1}, 5.5) &= 0.00\n",
    "\\end{aligned}\n",
    "\\end{equation*}\n",
    "\n",
    "\\begin{equation*}\n",
    "\\begin{aligned}\n",
    "    \\text{GiniGain}(\\text{Gene 1}, 5.7) &= \\left( \\frac{|D_l|}{|D_t|} \\text{Gini}(D_l) + \\frac{|D_r|}{|D_t|} \\text{Gini}(D_r) \\right)  &\n",
    "    \\text{GiniGain}(\\text{Gene 1}, 6.3) &= \\left( \\frac{|D_l|}{|D_t|} \\text{Gini}(D_l) + \\frac{|D_r|}{|D_t|} \\text{Gini}(D_r) \\right)  \\\\\n",
    "    \n",
    "                                        &= \\left( \\left( \\frac{7}{12}  * 0.24  \\right) + \\left( \\frac{5}{12}  * 0.00 \\right)   \\right)  &\n",
    "                                        &= \\left( \\left( \\frac{8}{12}  * 0.38  \\right) + \\left( \\frac{4}{12}  * 0.00 \\right)   \\right)  \\\\\n",
    "    \\text{GiniGain}(\\text{Gene 1}, 5.7) &= 0.14 & \n",
    "    \\text{GiniGain}(\\text{Gene 1}, 6.3) &= 0.25\n",
    "\\end{aligned}\n",
    "\\end{equation*}\n",
    "\n",
    "\n",
    "\\begin{equation*}\n",
    "\\begin{aligned}\n",
    "    \\text{GiniGain}(\\text{Gene 1}, 6.4) &= \\left( \\frac{|D_l|}{|D_t|} \\text{Gini}(D_l) + \\frac{|D_r|}{|D_t|} \\text{Gini}(D_r) \\right)  &\n",
    "    \\text{GiniGain}(\\text{Gene 1}, 6.5) &= \\left( \\frac{|D_l|}{|D_t|} \\text{Gini}(D_l) + \\frac{|D_r|}{|D_t|} \\text{Gini}(D_r) \\right)  \\\\\n",
    "    \n",
    "                                        &= \\left( \\left( \\frac{9}{12}  * 0.44  \\right) + \\left( \\frac{3}{12}  * 0.00 \\right)   \\right)  &\n",
    "                                        &= \\left( \\left( \\frac{10}{12} * 0.48  \\right) + \\left( \\frac{2}{12}  * 0.00 \\right)   \\right)  \\\\\n",
    "    \\text{GiniGain}(\\text{Gene 1}, 6.4) &= 0.33 & \n",
    "    \\text{GiniGain}(\\text{Gene 1}, 6.5) &= 0.40\n",
    "\\end{aligned}\n",
    "\\end{equation*}\n",
    "\n",
    "\n",
    "\\begin{equation*}\n",
    "\\begin{aligned}\n",
    "    \\text{GiniGain}(\\text{Gene 1}, 6.9) &= \\left( \\frac{|D_l|}{|D_t|} \\text{Gini}(D_l) + \\frac{|D_r|}{|D_t|} \\text{Gini}(D_r) \\right)  \\\\\n",
    "                                        &= \\left( \\left( \\frac{11}{12} * 0.50  \\right) + \\left( \\frac{1}{12}  * 0.00 \\right)   \\right)  \\\\\n",
    "    \\text{GiniGain}(\\text{Gene 1}, 6.9) &= 0.45\n",
    "\\end{aligned}\n",
    "\\end{equation*}\n",
    "\n",
    "\n",
    "Since the <code>weighted_gini</code> < <code>best_gini = 0.50</code> last occurred at value 5.4 before 5.5, we get the split value by finding the midpoint between these two values. That is, \n",
    "$$ \\text{split value} = \\frac{5.4 + 5.5}{2} = 5.45 $$\n",
    "\n",
    "Therefore, the possible split value for Gene 1 is 5.45 with gini impurity of 0.50."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's proceed to split the data based on Gene 2. Similarly, we first sort the values in ascending order:\n",
    "$$ [2.3, 2.8, 2.8, 3.0, 3.1, 3.1, 3.2, 3.3, 3.4, 3.4, 3.4, 3.5] $$\n",
    "Then, let's calculate gini index and gini gain of split values of attributes using the same procedure in Gene 1.\n",
    "\n",
    "\\begin{equation*}\n",
    "\\begin{aligned}\n",
    "    \\text{GiniGain}(\\text{Gene 2}, 2.8) &= \\left( \\frac{|D_l|}{|D_t|} \\text{Gini}(D_l) + \\frac{|D_r|}{|D_t|} \\text{Gini}(D_r) \\right)  &\n",
    "    \\text{GiniGain}(\\text{Gene 2}, 3.0) &= \\left( \\frac{|D_l|}{|D_t|} \\text{Gini}(D_l) + \\frac{|D_r|}{|D_t|} \\text{Gini}(D_r) \\right)  \\\\\n",
    "    \n",
    "                                        &= \\left( \\left( \\frac{1}{12}  * 0.00    \\right) + \\left( \\frac{11}{12} * 0.50 \\right)   \\right)  &\n",
    "                                        &= \\left( \\left( \\frac{3}{12}  * 0.00    \\right) + \\left( \\frac{9}{12}  * 0.44 \\right)   \\right)  \\\\\n",
    "    \\text{GiniGain}(\\text{Gene 2}, 2.8) &= 0.45 & \n",
    "    \\text{GiniGain}(\\text{Gene 2}, 3.0) &= 0.33\n",
    "\\end{aligned}\n",
    "\\end{equation*}\n",
    "\n",
    "\\begin{equation*}\n",
    "\\begin{aligned}\n",
    "    \\text{GiniGain}(\\text{Gene 2}, 3.1) &= \\left( \\frac{|D_l|}{|D_t|} \\text{Gini}(D_l) + \\frac{|D_r|}{|D_t|} \\text{Gini}(D_r) \\right)  &\n",
    "    \\text{GiniGain}(\\text{Gene 2}, 3.2) &= \\left( \\frac{|D_l|}{|D_t|} \\text{Gini}(D_l) + \\frac{|D_r|}{|D_t|} \\text{Gini}(D_r) \\right)  \\\\\n",
    "    \n",
    "                                        &= \\left( \\left( \\frac{4}{12}  * 0.38    \\right) + \\left( \\frac{8}{12}  * 0.47 \\right)   \\right)  &\n",
    "                                        &= \\left( \\left( \\frac{6}{12}  * 0.44    \\right) + \\left( \\frac{6}{12}  * 0.44 \\right)   \\right)  \\\\\n",
    "    \\text{GiniGain}(\\text{Gene 2}, 3.1) &= 0.44 & \n",
    "    \\text{GiniGain}(\\text{Gene 2}, 3.2) &= 0.44\n",
    "\\end{aligned}\n",
    "\\end{equation*}\n",
    "\n",
    "\\begin{equation*}\n",
    "\\begin{aligned}\n",
    "    \\text{GiniGain}(\\text{Gene 2}, 3.3) &= \\left( \\frac{|D_l|}{|D_t|} \\text{Gini}(D_l) + \\frac{|D_r|}{|D_t|} \\text{Gini}(D_r) \\right)  &\n",
    "    \\text{GiniGain}(\\text{Gene 2}, 3.4) &= \\left( \\frac{|D_l|}{|D_t|} \\text{Gini}(D_l) + \\frac{|D_r|}{|D_t|} \\text{Gini}(D_r) \\right)  \\\\\n",
    "    \n",
    "                                        &= \\left( \\left( \\frac{7}{12}  * 0.41    \\right) + \\left( \\frac{5}{12}  * 0.32 \\right)   \\right)  &\n",
    "                                        &= \\left( \\left( \\frac{8}{12}  * 0.38    \\right) + \\left( \\frac{4}{12}  * 0.00 \\right)   \\right)  \\\\\n",
    "    \\text{GiniGain}(\\text{Gene 2}, 3.3) &= 0.37 & \n",
    "    \\text{GiniGain}(\\text{Gene 2}, 3.4) &= 0.25\n",
    "\\end{aligned}\n",
    "\\end{equation*}\n",
    "\n",
    "\\begin{equation*}\n",
    "\\begin{aligned}\n",
    "    \\text{GiniGain}(\\text{Gene 1}, 3.5) &= \\left( \\frac{|D_l|}{|D_t|} \\text{Gini}(D_l) + \\frac{|D_r|}{|D_t|} \\text{Gini}(D_r) \\right)  \\\\\n",
    "                                        &= \\left( \\left( \\frac{11}{12} * 0.50  \\right) + \\left( \\frac{1}{12}  * 0.00 \\right)   \\right)  \\\\\n",
    "    \\text{GiniGain}(\\text{Gene 1}, 3.5) &= 0.45\n",
    "\\end{aligned}\n",
    "\\end{equation*}\n",
    "\n",
    "If **chosen**, the possible split value for Gene 2 is 3.35 given that its <code>weighted_gini: 0.25 < best_gini</code> such that \n",
    "$$ \\text{split value} = \\frac{3.30 + 3.40}{2} = 3.35 $$"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, let us split the data based on Gene 3. We will use the same procedure from Gene 1 and Gene 2.\n",
    "$$ [1.5, 1.5, 1.6, 1.6, 1.6, 1.9, 4.0, 4.5, 4.6, 4.7, 4.7, 4.9] $$\n",
    "Then, let's calculate gini index and gini gain of split values of attributes.\n",
    "\n",
    "\\begin{equation*}\n",
    "\\begin{aligned}\n",
    "    \\text{GiniGain}(\\text{Gene 3}, 1.6) &= \\left( \\frac{|D_l|}{|D_t|} \\text{Gini}(D_l) + \\frac{|D_r|}{|D_t|} \\text{Gini}(D_r) \\right)  &\n",
    "    \\text{GiniGain}(\\text{Gene 3}, 1.9) &= \\left( \\frac{|D_l|}{|D_t|} \\text{Gini}(D_l) + \\frac{|D_r|}{|D_t|} \\text{Gini}(D_r) \\right)  \\\\\n",
    "    \n",
    "                                        &= \\left( \\left( \\frac{2}{12}  * 0.00    \\right) + \\left( \\frac{10}{12} * 0.48 \\right)   \\right)  &\n",
    "                                        &= \\left( \\left( \\frac{5}{12}  * 0.00    \\right) + \\left( \\frac{7}{12}  * 0.24 \\right)   \\right)  \\\\\n",
    "    \\text{GiniGain}(\\text{Gene 3}, 1.6) &= 0.40 & \n",
    "    \\text{GiniGain}(\\text{Gene 3}, 1.9) &= 0.14\n",
    "\\end{aligned}\n",
    "\\end{equation*}\n",
    "\n",
    "\n",
    "\\begin{equation*}\n",
    "\\begin{aligned}\n",
    "    \\text{GiniGain}(\\text{Gene 3}, 4.0) &= \\left( \\frac{|D_l|}{|D_t|} \\text{Gini}(D_l) + \\frac{|D_r|}{|D_t|} \\text{Gini}(D_r) \\right)  &\n",
    "    \\text{GiniGain}(\\text{Gene 3}, 4.5) &= \\left( \\frac{|D_l|}{|D_t|} \\text{Gini}(D_l) + \\frac{|D_r|}{|D_t|} \\text{Gini}(D_r) \\right)  \\\\\n",
    "    \n",
    "                                        &= \\left( \\left( \\frac{6}{12}  * 0.00    \\right) + \\left( \\frac{6}{12}  * 0.00 \\right)   \\right)  &\n",
    "                                        &= \\left( \\left( \\frac{7}{12}  * 0.24    \\right) + \\left( \\frac{5}{12}  * 0.00 \\right)   \\right)  \\\\\n",
    "    \\text{GiniGain}(\\text{Gene 3}, 4.0) &= 0.00 & \n",
    "    \\text{GiniGain}(\\text{Gene 3}, 4.5) &= 0.14\n",
    "\\end{aligned}\n",
    "\\end{equation*}\n",
    "\n",
    "\n",
    "\\begin{equation*}\n",
    "\\begin{aligned}\n",
    "    \\text{GiniGain}(\\text{Gene 3}, 4.6) &= \\left( \\frac{|D_l|}{|D_t|} \\text{Gini}(D_l) + \\frac{|D_r|}{|D_t|} \\text{Gini}(D_r) \\right)  &\n",
    "    \\text{GiniGain}(\\text{Gene 3}, 4.7) &= \\left( \\frac{|D_l|}{|D_t|} \\text{Gini}(D_l) + \\frac{|D_r|}{|D_t|} \\text{Gini}(D_r) \\right)  \\\\\n",
    "    \n",
    "                                        &= \\left( \\left( \\frac{8}{12}  * 0.38    \\right) + \\left( \\frac{4}{12}  * 0.00 \\right)   \\right)  &\n",
    "                                        &= \\left( \\left( \\frac{9}{12}  * 0.44    \\right) + \\left( \\frac{3}{12}  * 0.00 \\right)   \\right)  \\\\\n",
    "    \\text{GiniGain}(\\text{Gene 3}, 4.6) &= 0.25 & \n",
    "    \\text{GiniGain}(\\text{Gene 3}, 4.7) &= 0.33\n",
    "\\end{aligned}\n",
    "\\end{equation*}\n",
    "\n",
    "\\begin{equation*}\n",
    "\\begin{aligned}\n",
    "    \\text{GiniGain}(\\text{Gene 1}, 4.9) &= \\left( \\frac{|D_l|}{|D_t|} \\text{Gini}(D_l) + \\frac{|D_r|}{|D_t|} \\text{Gini}(D_r) \\right)  \\\\\n",
    "                                        &= \\left( \\left( \\frac{11}{12} * 0.50  \\right) + \\left( \\frac{1}{12}  * 0.00 \\right)   \\right)  \\\\\n",
    "    \\text{GiniGain}(\\text{Gene 1}, 4.9) &= 0.45\n",
    "\\end{aligned}\n",
    "\\end{equation*}\n",
    "\n",
    "Since the <code>weighted_gini</code> < <code>best_gini</code> last occurred at value 1.9 before 4.0, we get the split value by finding the midpoint between these two values. That is, \n",
    "$$ \\text{split value} = \\frac{1.90 + 4.0}{2} = 2.95 $$\n",
    "\n",
    "Therefore, the possible split value for Gene 3 is 2.95 with gini impurity of 0.50."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, let us split the data based on Gene 4 using the sorted values:\n",
    "$$ [0.2, 0.2, 0.2, 0.2, 0.4, 1.2, 1.3, 1.3, 1.5, 1.5, 1.5, 1.6] $$\n",
    "\n",
    "\\begin{equation*}\n",
    "\\begin{aligned}\n",
    "    \\text{GiniGain}(\\text{Gene 4}, 0.4) &= \\left( \\frac{|D_l|}{|D_t|} \\text{Gini}(D_l) + \\frac{|D_r|}{|D_t|} \\text{Gini}(D_r) \\right)  &\n",
    "    \\text{GiniGain}(\\text{Gene 4}, 1.2) &= \\left( \\frac{|D_l|}{|D_t|} \\text{Gini}(D_l) + \\frac{|D_r|}{|D_t|} \\text{Gini}(D_r) \\right)  \\\\\n",
    "    \n",
    "                                        &= \\left( \\left( \\frac{4}{12}  * 0.00    \\right) + \\left( \\frac{8}{12}  * 0.38 \\right)   \\right)  &\n",
    "                                        &= \\left( \\left( \\frac{5}{12}  * 0.00    \\right) + \\left( \\frac{7}{12}  * 0.24 \\right)   \\right)  \\\\\n",
    "    \\text{GiniGain}(\\text{Gene 4}, 0.4) &= 0.25 & \n",
    "    \\text{GiniGain}(\\text{Gene 4}, 1.2) &= 0.14\n",
    "\\end{aligned}\n",
    "\\end{equation*}\n",
    "\n",
    "\n",
    "\\begin{equation*}\n",
    "\\begin{aligned}\n",
    "    \\text{GiniGain}(\\text{Gene 4}, 1.3) &= \\left( \\frac{|D_l|}{|D_t|} \\text{Gini}(D_l) + \\frac{|D_r|}{|D_t|} \\text{Gini}(D_r) \\right)  &\n",
    "    \\text{GiniGain}(\\text{Gene 4}, 1.5) &= \\left( \\frac{|D_l|}{|D_t|} \\text{Gini}(D_l) + \\frac{|D_r|}{|D_t|} \\text{Gini}(D_r) \\right)  \\\\\n",
    "    \n",
    "                                        &= \\left( \\left( \\frac{6}{12}  * 0.00    \\right) + \\left( \\frac{6}{12}  * 0.00 \\right)   \\right)  &\n",
    "                                        &= \\left( \\left( \\frac{8}{12}  * 0.38    \\right) + \\left( \\frac{4}{12}  * 0.00 \\right)   \\right)  \\\\\n",
    "    \\text{GiniGain}(\\text{Gene 4}, 1.3) &= 0.00 & \n",
    "    \\text{GiniGain}(\\text{Gene 4}, 1.5) &= 0.25\n",
    "\\end{aligned}\n",
    "\\end{equation*}\n",
    "\n",
    "\n",
    "\\begin{equation*}\n",
    "\\begin{aligned}\n",
    "    \\text{GiniGain}(\\text{Gene 4}, 1.6) &= \\left( \\frac{|D_l|}{|D_t|} \\text{Gini}(D_l) + \\frac{|D_r|}{|D_t|} \\text{Gini}(D_r) \\right)  \\\\\n",
    "                                        &= \\left( \\left( \\frac{11}{12} * 0.50  \\right) + \\left( \\frac{1}{12}  * 0.00 \\right)   \\right)  \\\\\n",
    "    \\text{GiniGain}(\\text{Gene 4}, 1.6) &= 0.45\n",
    "\\end{aligned}\n",
    "\\end{equation*}\n",
    "\n",
    "Since the <code>weighted_gini</code> < <code>best_gini = 0.50</code> last occurred at value 1.2 before 1.3, we get the split value by finding the midpoint between these two values. That is, \n",
    "$$ \\text{split value} = \\frac{1.2 + 1.3}{2} = 1.25 $$\n",
    "\n",
    "Therefore, the possible split value for Gene 4 is 1.25 with gini impurity of 0.50."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The features Gene 1, Gene 3, and Gene 4 have similar gini impurity values. By theory, we can utilize any feature selection methods to choose among these three feature to build the root node. In this assignment, we jsut randomly choose Gene 4 with split value at 1.25 and gini impurity of 0.50.\n",
    "\n",
    "<center>\n",
    "<img src = \"../figures/tree.PNG\" width = \"750\"/> <br>\n",
    "</center>"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, let's calculate the <code>log(odds)</code> based on the residual and previous probability from the table above. The most common transformation formula in Gradient Boosting for Classification is as follows:\n",
    "\n",
    "\\begin{equation*}\n",
    "\\begin{aligned}\n",
    "    \\text{transform} = \\frac{\\sum r_i}{\\sum \\left( p_i \\cdot (1 - p_i) \\right)}\n",
    "\\end{aligned}\n",
    "\\end{equation*}\n",
    "\n",
    "\n",
    "\\begin{equation*}\n",
    "\\begin{aligned}\n",
    "    \\text{transform}(0.50)  &= \\frac{0.50 + 0.50 + 0.50 + 0.50 + 0.50 +0.50}{\\sum \\left( 0.50 \\cdot (1 - 0.50) \\right)}  &\n",
    "\n",
    "    \\text{transform}-(0.50) &= \\frac{-0.50 - 0.50 - 0.50 - 0.50 - 0.50 - 0.50}{\\sum \\left(0.50 \\cdot (1 - 0.50) \\right)} \\\\\n",
    "\n",
    "                            &= \\frac{6 \\cdot ( 0.50)}{6 \\left( 0.50 \\cdot (1 - 0.50) \\right)} &\n",
    "                            &= \\frac{6 \\cdot (-0.50)}{6 \\left( 0.50 \\cdot (1 - 0.50) \\right)} \\\\\n",
    "    \\text{transform}(0.50)  &= 2 &\n",
    "    \\text{transform}(-0.50)  &= -2\n",
    "\\end{aligned}\n",
    "\\end{equation*}\n",
    "\n",
    "From the transformation, the <code>log(odds)</code> of having cancer  and not having cancer are:\n",
    "\\begin{equation*}\n",
    "\\begin{aligned}\n",
    "                    \\text{log(odds)(Cancer)}    = \\frac{e^{0.6}}{1 + e^{0.6}}   = 0.65 & & & & & & &\n",
    "                    \\text{log(odds)(No Cancer)} = \\frac{e^{-0.6}}{1 + e^{-0.6}} = 0.35 \n",
    "\\end{aligned}\n",
    "\\end{equation*}\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can then calculate the prediction and the residual using the decision tree.\n",
    "<center>\n",
    "<img src = \"../figures/fig2.PNG\" width = \"850\"/> <br>\n",
    "</center>"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Therefore, the final Gradient Boosting tree ($M = 1$) is depicted as follows:\n",
    "\n",
    "<center>\n",
    "<img src = \"../figures/final.PNG\" width = \"850\"/> <br>\n",
    "</center>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "3608206ce1eb198bd23abae205dd191f991de1c92dbe872a18ef9e948d8a869d"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
